src/snowflake/snowpark/_internal/data_source/jdbc.py

```diff
@@ -0,0 +1,582 @@
+    TINYINT = "TINYINT"
+    VARBINARY = "VARBINARY"
+    VARCHAR = "VARCHAR"
+    NOT_SUPPORTED = "NOT_SUPPORTED"
+    NONE = None
+
+
+JAVA_TYPE_TO_SNOWFLAKE_TYPE = {
+    "java.math.BigDecimal": DecimalType,
+    "java.math.BigInteger": DecimalType,
+    "java.lang.Float": FloatType,
+    "java.lang.Double": FloatType,
+    "java.lang.Boolean": BooleanType,
+    "java.lang.String": StringType,
+    "java.sql.Timestamp": TimestampType,
+    "java.sql.Date": DateType,
+    "java.sql.Array": ArrayType,
+    "java.lang.String[]": ArrayType,
+    "java.lang.Long": DecimalType,
+    "java.lang.Integer": DecimalType,
+    "java.lang.Short": DecimalType,
+}
+JDBC_TYPE_TO_SNOWFLAKE_TYPE = {
+    JDBCType.BIGINT: IntegerType,
+    JDBCType.BINARY: StringType,
+    JDBCType.BIT: StringType,
+    JDBCType.BOOLEAN: BooleanType,
+    JDBCType.CHAR: StringType,
+    JDBCType.CLOB: StringType,
+    JDBCType.DATE: DateType,
+    JDBCType.DECIMAL: DecimalType,
+    JDBCType.DOUBLE: DoubleType,
+    JDBCType.FLOAT: FloatType,
+    JDBCType.INTEGER: IntegerType,
+    JDBCType.LONGNVARCHAR: StringType,
+    JDBCType.LONGVARBINARY: StringType,
+    JDBCType.LONGVARCHAR: StringType,
+    JDBCType.NCHAR: StringType,
+    JDBCType.NCLOB: StringType,
+    JDBCType.NUMERIC: DecimalType,
+    JDBCType.NVARCHAR: StringType,
+    JDBCType.OTHER: StringType,
+    JDBCType.SMALLINT: IntegerType,
+    JDBCType.TIME: TimeType,
+    JDBCType.TIME_WITH_TIMEZONE: TimestampType,
+    JDBCType.TIMESTAMP: TimestampType,
+    JDBCType.TIMESTAMP_WITH_TIMEZONE: TimestampType,
+    JDBCType.TINYINT: IntegerType,
+    JDBCType.VARBINARY: StringType,
+    JDBCType.VARCHAR: StringType,
+}
+
+PARTITION_TABLE_COLUMN_NAME = "partition"
+
+
+class JDBC:
+    def __init__(
+        self,
+        session: "Session",
+        url: str,
+        table_or_query: str,
+        external_access_integration: str,
+        imports: List[str],
+        is_query: bool,
+        secret: str,
+        *,
+        properties: Optional[dict] = None,
+        packages: Optional[List[str]] = None,
+        java_version: Optional[int] = 17,
+        column: Optional[str] = None,
+        lower_bound: Optional[Union[str, int]] = None,
+        upper_bound: Optional[Union[str, int]] = None,
+        num_partitions: Optional[int] = None,
+        query_timeout: Optional[int] = 0,
+        fetch_size: Optional[int] = 0,
+        custom_schema: Optional[Union[str, StructType]] = None,
+        predicates: Optional[List[str]] = None,
+        session_init_statement: Optional[List[str]] = None,
+        _emit_ast: bool = True,
+    ) -> None:
+        self.session = session
+        self.url = url
+        self.table_or_query = table_or_query
+
+        self.external_access_integration = external_access_integration
+        self.imports = imports
+        self.packages = packages
+        self.java_version = java_version
+        self.secret = secret
+        self.properties = properties
```

Comment:
let's make a copy the properties dict, otherwise it's a reference to the original user input.
code like below alters user's original dict.

```python
                    del self.properties[key]
```